src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs

```diff
@@ -874,11 +873,24 @@ public string ProcessName
                     else
                     {
                         // If we don't have a populated ProcessInfo, then get and cache the process name.
-                        _processName = ProcessManager.GetProcessName(_processId, _machineName);
+                        try
+                        {
+                            // Ensure that the process is not yet exited
+                            EnsureState(State.HaveNonExitedId);
+                            _processName = ProcessManager.GetProcessName(_processId, _machineName);
 
-                        if (_processName == null)
+                            // Ensure that a process name is available
+                            if (_processName == null)
+                            {
+                                throw new InvalidOperationException(SR.NoProcessInfo);
+                            }
+                        }
+                        catch
```

Comment:
Might it be worth narrowing down the exception that we expect?

Reply:
It's much easier - as `ProcessManager.GetProcessName` simply returns null in case it does not get a valid result from the kernel32 call, it is not neccessary to throw an exception and then catch it. I moved the fallback code to line 885.